src/OpenTelemetry/Internal/CircularBuffer.cs

```diff
@@ -83,22 +86,23 @@ public bool Add(T value)
 
             while (true)
             {
-                var tailSnapshot = this.tail;
-                var headSnapshot = this.head;
+                var tailSnapshot = Volatile.Read(ref this.tail);
+                var headSnapshot = Volatile.Read(ref this.head);
 
                 if (headSnapshot - tailSnapshot >= this.Capacity)
                 {
                     return false; // buffer is full
                 }
 
-                var head = Interlocked.CompareExchange(ref this.head, headSnapshot + 1, headSnapshot);
-                if (head != headSnapshot)
+                if (Interlocked.CompareExchange(ref this.head, headSnapshot + 1, headSnapshot) != headSnapshot)
                 {
                     continue;
```

Comment:
Would it be useful to add ` Spinwait.SpinOnce()` here before `continue`?

Reply:
Defer to @reyang on this one, but I don't think so. This thread isn't so much waiting on another thread to finish as it is learning that some other thread took the head. It should retry immediately and just take the next head/index available.

Reply:
SpinOnce might be more smart yielding if singlecore etc?

Reply:
I think a yield here would be even worse than a spin 😄 Because it doesn't need to wait on anything. If there is only 1 core this thread should just loop around, take the next index, and continue on doing its thing. Same as if there were many cores, really.

There should probably be a SpinOnce here though: https://github.com/open-telemetry/opentelemetry-dotnet/blob/91d7e83f232f5c9ab01b9610fb4597b24fb48955/src/OpenTelemetry/Internal/CircularBuffer.cs#L171

Because that logic is actually waiting on the writer thread to finish. On single core, it should yield immediately because a spin won't accomplish anything other than delay letting the writer get the CPU back to finish its job 🤣

Reply:
> I think a yield here would be even worse than a spin

+1